io/sam/sam.go

```diff
@@ -0,0 +1,202 @@
+
+// NewParser creates a parser from an io.Reader for sam data. For larger
+// alignments, you will want to increase the maxLineSize.
+func NewParser(r io.Reader, maxLineSize int) (*Parser, Header, error) {
+	parser := &Parser{
+		reader: *bufio.NewReaderSize(r, maxLineSize),
+	}
+	var header Header
+	var hdParsed bool
+	// Initialize header maps
+	header.HD = make(map[string]string)
+	header.SQ = []map[string]string{}
+	header.RG = []map[string]string{}
+	header.PG = []map[string]string{}
+	header.CO = []string{}
+
+	// We need to first read the header before returning the parser to the
+	// user for analyzing alignments.
+	for {
+		lineBytes, err := parser.reader.ReadSlice('\n')
+		if err != nil {
+			return parser, Header{}, err
+		}
+		line := strings.TrimSpace(string(lineBytes))
+		parser.line++
+		if len(line) == 0 {
+			return parser, Header{}, fmt.Errorf("Line %d is empty. Empty lines are not allowed in headers.", parser.line)
+		}
+		// If this line is the start of the alignments, set the firstLine
+		// into memory, and then break this loop.
+		if line[0] != '@' {
+			parser.firstLine = line
+			break
+		}
+		values := strings.Split(line, "\t")
+		if len(values) < 1 {
+			return parser, Header{}, fmt.Errorf("Line %d should contain at least 1 value. Got: %d. Line text: %s", parser.line, len(values), line)
+		}
+
+		// If we haven't parsed HD, it is always the first line: lets parse it.
+		if !hdParsed {
+			if values[0] != "@HD" {
+				return parser, Header{}, fmt.Errorf("First line (%d) should always contain @HD first. Line text: %s", parser.line, line)
+			}
+			// Now parse the rest of the HD header
+			for _, value := range values[1:] {
+				valueSplit := strings.Split(value, ":")
+				header.HD[valueSplit[0]] = valueSplit[1]
+			}
+			hdParsed = true
+			continue
+		}
+
+		// CO lines are unique in that they are just strings. So we try to parse them
+		// first. We include the entire comment line for these.
+		if values[0] == "@CO" {
+			header.CO = append(header.CO, line)
+			continue
+		}
+
+		// HD/CO lines have been successfully parsed, now we work on SQ, RG, and PG.
+		// Luckily, each one has an identical form ( TAG:DATA ), so we can parse that
+		// first and then just apply it to the respect top level tag.
+		genericMap := make(map[string]string)
+		for _, value := range values[1:] {
+			valueSplit := strings.Split(value, ":")
+			genericMap[valueSplit[0]] = valueSplit[1]
+		}
+		switch values[0] {
+		case "@SQ":
+			header.SQ = append(header.SQ, genericMap)
+		case "@RG":
+			header.RG = append(header.RG, genericMap)
+		case "@PG":
+			header.PG = append(header.PG, genericMap)
+		default:
+			return parser, Header{}, fmt.Errorf("Line %d should contain @SQ, @RG, @PG or @CO as top level tags, but they weren't found. Line text: %s", parser.line, line)
+		}
+	}
+	return parser, header, nil
+}
+
+// ParseNext parsers the next read from a parser. Returns an error upon EOF.
+func (parser *Parser) ParseNext() (Alignment, error) {
```

Comment:
Please add some whitespace/comments to this to make it easier to read!

Reply:
Ah yes, I definitely need to do that. Lil hard to read right now.